File: relationship-manager/conversation-service/src/main/java/com/bank/rm/conversation/ai/EntityExtractor.java
package com.bank.rm.conversation.ai;

import com.bank.rm.common.dto.ConversationPhase;
import dev.langchain4j.model.chat.ChatLanguageModel;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.HashMap;
import java.util.Map;

public class EntityExtractor {
    private static final Logger log = LoggerFactory.getLogger(EntityExtractor.class);
    private final ChatLanguageModel chatModel;
    private final ObjectMapper objectMapper;

    public EntityExtractor(ChatLanguageModel chatModel, ObjectMapper objectMapper) {
        this.chatModel = chatModel;
        this.objectMapper = objectMapper;
    }

    @SuppressWarnings("unchecked")
    public Map<String, Object> extract(String userMessage, ConversationPhase phase,
                                        Map<String, Object> existingEntities) {
        Map<String, Object> entities = new HashMap<>(existingEntities);
        try {
            String prompt = String.format(
                "Extract structured data from this customer message in a banking conversation.\n" +
                "Current phase: %s\nCustomer said: \"%s\"\nAlready collected: %s\n\n" +
                "Return ONLY a JSON object with extracted key-value pairs. " +
                "Keys should be: name, ageGroup, location, phone, email, incomeSource, incomeRange, " +
                "savings, investments, retirementTarget, retirementAge, dependents, riskTolerance, " +
                "preferredChannel, followUpFrequency. Only include keys found in the message.",
                phase, userMessage, objectMapper.writeValueAsString(existingEntities));
            String response = chatModel.generate(prompt);
            String jsonStr = response.contains("{") ?
                    response.substring(response.indexOf("{"), response.lastIndexOf("}") + 1) : "{}";
            Map<String, Object> extracted = objectMapper.readValue(jsonStr, Map.class);
            entities.putAll(extracted);
        } catch (Exception e) {
            log.warn("LLM entity extraction failed, using regex fallback: {}", e.getMessage());
            entities.putAll(regexFallback(userMessage, phase));
        }
        return entities;
    }

    private Map<String, Object> regexFallback(String message, ConversationPhase phase) {
        Map<String, Object> result = new HashMap<>();
        String lower = message.toLowerCase();
        if (phase == ConversationPhase.GREETING && message.length() < 100) {
            String cleaned = message.replaceAll("[^a-zA-Z\\s]", "").trim();
            String[] words = cleaned.split("\\s+");
            if (words.length > 0 && words.length <= 4) {
                result.put("name", cleaned);
            }
        }
        if (lower.contains("20-30") || lower.contains("twenties")) result.put("ageGroup", "AGE_20_30");
        if (lower.contains("30-40") || lower.contains("thirties")) result.put("ageGroup", "AGE_30_40");
        if (lower.contains("40-50") || lower.contains("forties")) result.put("ageGroup", "AGE_40_50");
        if (lower.contains("50+") || lower.contains("fifties")) result.put("ageGroup", "AGE_50_PLUS");
        if (lower.contains("salaried")) result.put("incomeSource", "SALARIED");
        if (lower.contains("business")) result.put("incomeSource", "BUSINESS");
        if (lower.contains("retired")) result.put("incomeSource", "RETIRED");
        if (lower.contains("whatsapp")) result.put("preferredChannel", "WHATSAPP");
        if (lower.contains("sms")) result.put("preferredChannel", "SMS");
        if (lower.contains("email") || lower.contains("mail")) result.put("preferredChannel", "EMAIL");
        if (lower.contains("phone") || lower.contains("call")) result.put("preferredChannel", "PHONE");
        if (lower.contains("weekly")) result.put("followUpFrequency", "WEEKLY");
        if (lower.contains("monthly")) result.put("followUpFrequency", "MONTHLY");
        if (lower.contains("quarterly")) result.put("followUpFrequency", "QUARTERLY");
        return result;
    }
}

File: relationship-manager/conversation-service/src/main/kotlin/com/bank/rm/conversation/ai/EntityExtractor.kt
package com.bank.rm.conversation.ai

import com.bank.rm.common.dto.ConversationPhase
import dev.langchain4j.model.chat.ChatLanguageModel
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class EntityExtractor(
    private val chatModel: ChatLanguageModel,
    private val objectMapper: ObjectMapper
) {
    private val log = LoggerFactory.getLogger(javaClass)

    // Regex patterns for quick extraction before LLM fallback
    private val patterns = mapOf(
        "name" to Regex("""(?:my name is|i'm|i am|call me)\s+([A-Z][a-z]+(?:\s+[A-Z][a-z]+)*)""", RegexOption.IGNORE_CASE),
        "age" to Regex("""(?:i'm|i am|age is)\s+(\d{2})""", RegexOption.IGNORE_CASE),
        "ageGroup" to Regex("""(20[s\-]30|30[s\-]40|40[s\-]50|50\+|fifty plus|twenties|thirties|forties|fifties)""", RegexOption.IGNORE_CASE),
        "email" to Regex("""[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"""),
        "phone" to Regex("""(?:\+91[\s-]?)?[6-9]\d{9}"""),
        "income" to Regex("""(\d+(?:\.\d+)?)\s*(?:lakh|lac|L|k|K|lpa|per\s*(?:annum|month|year))""", RegexOption.IGNORE_CASE),
        "incomeSource" to Regex("""(?:i (?:am|work as)|working as|profession is)\s+(salaried|business|professional|self[- ]employed|student|retired)""", RegexOption.IGNORE_CASE),
        "retirementAmount" to Regex("""(\d+(?:\.\d+)?)\s*(?:crore|cr|lakh|lac)\s*(?:by|at|when|for retirement)""", RegexOption.IGNORE_CASE)
    )

    fun extract(message: String, phase: ConversationPhase): Map<String, Any> {
        val quickExtracted = quickExtract(message, phase)
        if (quickExtracted.isNotEmpty()) return quickExtracted

        return llmExtract(message, phase)
    }

    private fun quickExtract(message: String, phase: ConversationPhase): Map<String, Any> {
        val result = mutableMapOf<String, Any>()

        when (phase) {
            ConversationPhase.GREETING -> {
                if (message.lowercase().let { it.contains("hi") || it.contains("hello") || it.contains("yes") || it.contains("sure") }) {
                    result["acknowledged"] = true
                }
            }
            ConversationPhase.PERSONAL -> {
                patterns["name"]?.find(message)?.groupValues?.get(1)?.let { result["name"] = it }
                patterns["email"]?.find(message)?.value?.let { result["email"] = it }
                patterns["phone"]?.find(message)?.value?.let { result["phone"] = it }
                extractAgeGroup(message)?.let { result["ageGroup"] = it }
            }
            ConversationPhase.FINANCIAL -> {
                patterns["incomeSource"]?.find(message)?.groupValues?.get(1)?.let {
                    result["incomeSource"] = normalizeIncomeSource(it)
                }
                extractIncomeRange(message)?.let { result["incomeRange"] = it }
            }
            ConversationPhase.GOALS -> {
                patterns["retirementAmount"]?.find(message)?.let {
                    result["retirementTarget"] = it.value
                }
                Regex("""retire\s*(?:by|at|around)\s*(\d{2})""", RegexOption.IGNORE_CASE).find(message)?.groupValues?.get(1)?.let {
                    result["retirementAge"] = it.toInt()
                }
            }
            ConversationPhase.CHANNEL_PREF -> {
                val channelKeywords = mapOf(
                    "sms" to "SMS", "text" to "SMS",
                    "whatsapp" to "WHATSAPP", "wa" to "WHATSAPP",
                    "email" to "EMAIL", "mail" to "EMAIL",
                    "phone" to "PHONE", "call" to "PHONE"
                )
                channelKeywords.entries.find { message.lowercase().contains(it.key) }?.let {
                    result["preferredChannel"] = it.value
                }
            }
            ConversationPhase.FOLLOWUP_SCHEDULE -> {
                val freqKeywords = mapOf(
                    "weekly" to "WEEKLY", "every week" to "WEEKLY",
                    "biweekly" to "BIWEEKLY", "every two weeks" to "BIWEEKLY", "fortnightly" to "BIWEEKLY",
                    "monthly" to "MONTHLY", "every month" to "MONTHLY",
                    "quarterly" to "QUARTERLY", "every three months" to "QUARTERLY", "every quarter" to "QUARTERLY"
                )
                freqKeywords.entries.find { message.lowercase().contains(it.key) }?.let {
                    result["followUpFrequency"] = it.value
                }
            }
            else -> {}
        }

        return result
    }

    private fun llmExtract(message: String, phase: ConversationPhase): Map<String, Any> {
        return try {
            val fields = when (phase) {
                ConversationPhase.PERSONAL -> "name, ageGroup (20-30/30-40/40-50/50+), location, phone, email"
                ConversationPhase.FINANCIAL -> "incomeSource (SALARIED/BUSINESS/PROFESSIONAL/SELF_EMPLOYED/STUDENT/RETIRED), incomeRange (60-70K/70-80K/80-100K/100-150K/150K+), currentSavings"
                ConversationPhase.GOALS -> "retirementTarget (amount), retirementAge"
                ConversationPhase.CHANNEL_PREF -> "preferredChannel (SMS/WHATSAPP/EMAIL/PHONE)"
                ConversationPhase.FOLLOWUP_SCHEDULE -> "followUpFrequency (WEEKLY/BIWEEKLY/MONTHLY/QUARTERLY)"
                else -> return emptyMap()
            }

            val response = chatModel.generate(
                """Extract structured data from this customer message. Return ONLY valid JSON with these possible fields: $fields
If a field is not mentioned, omit it. Do not guess.
Message: "$message"
JSON:"""
            )

            val jsonStr = response.trim().let {
                if (it.startsWith("```")) it.lines().drop(1).dropLast(1).joinToString("\n") else it
            }
            objectMapper.readValue<Map<String, Any>>(jsonStr)
        } catch (e: Exception) {
            log.warn("LLM entity extraction failed for phase $phase: ${e.message}")
            emptyMap()
        }
    }

    private fun extractAgeGroup(message: String): String? {
        val lower = message.lowercase()
        return when {
            lower.contains("20") && lower.contains("30") -> "AGE_20_30"
            lower.contains("30") && lower.contains("40") -> "AGE_30_40"
            lower.contains("40") && lower.contains("50") -> "AGE_40_50"
            lower.contains("50+") || lower.contains("fifty") || lower.contains("above 50") -> "AGE_50_PLUS"
            lower.contains("twenties") -> "AGE_20_30"
            lower.contains("thirties") -> "AGE_30_40"
            lower.contains("forties") -> "AGE_40_50"
            else -> {
                patterns["age"]?.find(message)?.groupValues?.get(1)?.toIntOrNull()?.let { age ->
                    when {
                        age < 30 -> "AGE_20_30"
                        age < 40 -> "AGE_30_40"
                        age < 50 -> "AGE_40_50"
                        else -> "AGE_50_PLUS"
                    }
                }
            }
        }
    }

    private fun extractIncomeRange(message: String): String? {
        val amount = patterns["income"]?.find(message)?.groupValues?.get(1)?.toDoubleOrNull() ?: return null
        val lower = message.lowercase()
        val annualAmount = when {
            lower.contains("month") -> amount * 12
            lower.contains("lakh") || lower.contains("lac") || lower.contains("l") -> amount * 100000
            lower.contains("k") -> amount * 1000
            else -> amount
        }
        return when {
            annualAmount < 70000 -> "60-70K"
            annualAmount < 80000 -> "70-80K"
            annualAmount < 100000 -> "80-100K"
            annualAmount < 150000 -> "100-150K"
            else -> "150K+"
        }
    }

    private fun normalizeIncomeSource(raw: String): String = when (raw.lowercase().trim()) {
        "salaried" -> "SALARIED"
        "business" -> "BUSINESS"
        "professional" -> "PROFESSIONAL"
        "self-employed", "self employed", "selfemployed" -> "SELF_EMPLOYED"
        "student" -> "STUDENT"
        "retired" -> "RETIRED"
        else -> raw.uppercase()
    }
}
